### ISSUE TITLE
Allow customizing how Flow views are included in the automatic main menu

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#19127]
### Describe your motivation

An automatically populated main menu is introduced for Hilla in Vaadin 24.4 and we're planning to add similar functionality for Flow in an upcoming version. The logic for populating this menu should automatically integrate with the server-side router access control that is used by the application so that the menu by default only lists views that are accessible for the current user.

### Describe the solution you'd like

By default, Flow should make views available for the automatic menu under these conditions:
* If the application has any `NavigationAccessControl` instances registered as listeners on the UI, then the view class must satisfy the `checkerList` and `decisionResolver` of those instances.
* The view has an `@Menu` annotation as described in https://github.com/vaadin/flow/issues/19109.

In addition, the list of views should by default be made available to the client only based on this additional condition:
* The application has a Hilla main layout based on the FS router, i.e. a `views/@layout.tsx` (or `.jsx`) file exists.

While those are the defaults, the application should have a way of overriding those defaults. One suggestion for how this could be done would be along the lines of this:
```java
public class MyServiceInitListener implements VaadinServiceInitListener {
    @Override
    public void serviceInit(ServiceInitEvent serviceInit) {
      Router router = serviceInit.getSource().getRouter();

      router.setPopulateClientSideMenu(true);
      router.setMenuFilter(menuFilterContext -> {
        if (menuFilterContext.hasRole("ADMIN") || !menuFilterContext.getLocation().getPath().startsWith("a")) {
          return menuFilterContext.getMenuItem();
        } else {
          return null;
        }
      });
    };
  }
}
```
The callback returns a menu item object to use for the view, or `null` if the view should not be included in the menu.

`menuFilterContext` gives access to these points of data:
* The default menu item. The implementation can use it as is or use it as a foundation for creating a custom item.
* Convenient access to the same logic that the default implementation uses to delegate to `NavigationAccessControl`, e.g. as `menuFilterContext.checkNavigationAccessControl()`
* All the same API as in `NavigationContext` except for the helpers to create `AccessCheckResult` return values.

The property set by `setPopulateClientSideMenu` is a tri-state choice where `Boolean.TRUE` always makes the list of accessible menu items available to the client. `Boolean.FALSE` never makes the list available and `null` (which is the default) makes the choice based on whether there's client-side main layout defined with the FS router (i.e. `views/@layout.tsx`).

**Comment by knoobie** [comment]
> In addition, the list of views should by default be made available to the client only based on this additional condition:

Is this condition needed at all? Isn't it enough to check for the usage of `@Menu`? If this only works with the client side oriented approach anyway - or do you plan to create another version on the server side?

**Comment by Legioth** [comment]
We do indeed plan to introduce automatic main menus also for main layouts built in Flow (targeted for Vaadin 24.5 in June) and then `@Menu` would be used to customize those as well. It would be confusing to have two different menu annotations and verbose to require setting a property in each individual `@Menu` annotation .

That's why we instead go with this heuristic that we expect to work as expected for 99% of all cases and additionally provide explicit configuration functionality for the few remaining cases.

**Comment by mshabarov** [comment]
> The implementation can use it as is or use it as a foundation for creating a custom item.

This isn't clear enough to me. For the return value of the filter callback I'd expect a `boolean` value, that instructs whether an item from menu items list should be sent to the client. Why the menu item have to be returned if the implementation already has a reference to any menu items passed into the callback?

> e.g. as menuFilterContext.checkNavigationAccessControl()

Having only this method or `menuFilterContext.hasAccess()` method should be enough, i.e. to just know if a user have an access or not, why one would want to check a role, principal or other auth parameters that `NavigationContext` exposes?

**Comment by knoobie** [comment]
Does this also take into account to have multiple hierarchies and / or grouped menus? Something often seen especially with larger applications? Also support for icons; prefix suffix and so on is needed / has to be replicated to to match the sidenav.. (as well as non-vaadin icons).. so many things.. just the obvious question: is this really needed?

cc @simasch ; might be interesting for you

**Comment by Legioth** [comment]
> Why the menu item have to be returned if the implementation already has a reference to any menu items passed into the callback?

My thinking here is that this same filter API could just as well also be used to have a centralized place of customizing the menu items (especially in the case where the actual rendering is done in the client). The key to doing that is that you could return a customized entry as an alternative to returning the original one.

> why one would want to check a role, principal or other auth parameters that NavigationContext exposes?

One of the main points of customizing the menu filtering is for applications that have custom (or legacy) access control rules that are not reflected through `NavigationAccessControl`. We cannot know exactly what values would be used as the input for that custom logic but it's a fair assumption that it's similar to what's already used with `NavigationAccessControl`. At the same time, we also want to have a helper there to make it trivial to create a custom filter that does the same as the default one except that it also includes views that aren't explicitly marked with `@Menu`.

> Does this also take into account to have multiple hierarchies and / or grouped menus? [...] so many things.. just the obvious question: is this really needed?

We acknowledge that the automatic main menu is mostly a tool for making it easy to get started with a simple application whereas many complex real-world application would outgrow the basic functionality. That's why the actual menu implementation will be a simple loop in the default `MainLayout.java` file that is included in various starters. The application developer can easily replace that logic with their own more sophisticated automation or a hard-coded list of views.

**Comment by Legioth** [comment]
After giving this some additional though, I started thinking that we might only need the `setPopulateClientSideMenu` functionality for now.

We can add the filter customization logic later and we should consider doing that in a slightly more complicated way that also takes client-side views into account at the same time. It might be best to consider implementing that functionality after we have built basic functionality for the automatic server-side menu.
